### ISSUE TITLE
Type checking `select` works differently on `useQueries` compared to `useQuery`

### ISSUE THREAD

**Issue by jtannas** [TanStack/query#6556]
### Describe the bug

When using `useQuery`, TS correctly infers the argument type of the `select` query option to be the return type of the `queryFn` data.
Applying the same settings to `useQueries` it loses that link and changes it to `any`.
The linked TS sandbox shows the issue better than I can explain it though.

### Your minimal, reproducible example

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1WO1AGhQ21wE84BfOAMyghDgCIABGAQwDtVOBjANYB6KOg58YAWgCOOKOWYAoRXwg94IcgBEOnOAF44AbQCMxAEzEAzMQAsxAKzEAbAF1lKtbziay8gyRYcuQAFAiKcHCyFADS6OQAXMYA5OgAbuhcya6EEVHBAGJcSSEAlAYAfHAACgwgwBgAdKKoEAA2GSGaOpyluZEYbegSJRxQUOX6VWNQjbTAbTC4IcCVcKsApHDmBvqGAAx9ilSlyqrqPuR++KgBaEG4N2HReARJRnnhkZEv5HGJKXSmWy-W+vyKJUmVVqjAa6GaBHanW6ug4R2+cEGwxgo3GULgMzmCyWUBWa02212B3R1FyrhOiiAA

### Steps to reproduce

See linked Typescript playground to a minimal reproducible example

### Expected behavior

I would like the type checking of `select` on `useQueries` to work as well as it does on `useQuery`.

### How often does this bug happen?

Every time

### Platform

- Ubuntu 22.04

### Tanstack Query adapter

react-query

### TanStack Query version

v5.3.2

### TypeScript version

v5.14.0

**Comment by TkDodo** [comment]
I think it's a known limitation.

Possible workarounds:

- annotate the input to `select`: [playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1WO1AGhQ21wE84BfOAMyghDgCIABGAQwDtVOBjANYB6KOg58YAWgCOOKOWYAoRXwg94IcgBEOnOAF44AbQCMxAEzEAzMQAsxAKzEAbAF1lKtbziay8gyRYcuQAFAiKcHCyFADS6OQAXMYA5OgAbuhcya6EEVHBAGJcSSEAlAYAfHAACgwgwBgAdKKoEAA2GSGaOpyluZEYbegSJRxQUOX6VWNQjSAcYCEAyjB4XADmpYpUW57qPuR++KgBaEG4x2HReARJRnnhkZHX5HGJKemZ2f1PL0UlkyqtUYDXQzQI7U63V0HD6eQG6CGIzgIRmSQAguMOOQADxcZAgABGuAqgLgMzmC2Wq2AGy2kSouVcO0UQA)
- use the `queryOptions` helper: [playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1WO1AGhQ21wE9iBHHKcgeTBmAgDtU4BfOAMyghBwARAAEYAQ3YSAxgGsA9FHTjpMALQ0KQgFDbpbVPBDkAIuIlwAvHADaARmIAmYgGZiAFmIBWYgDYAurp6BkbkZHRWJFi05AAUCNpwcJp0ANLo5ABctgDk6ABu6Kw5-oSJyTEAYqzZsQCUVgB8cAAK-CDAGAB0SqgQADaFscZmEnVlSRj96Kq14lBQDZbN81BdIOJgsQDKMHisAOZ12pzHwVJwxuH4HNZo0bg38Sk32TblKQxMLOzx5UmfdJZXIFIolCZJCoUaq1JbNNoCTroHoEAZDEbmcTjf5wKYzGBzBZwuCrdabHZ7YCHY5JU5lfynbRAA)

**Comment by TkDodo** [comment]
I'll leave this open if someone wants to take a look, but given that we have good workarounds, it's not a priority for me.

**Comment by jtannas** [comment]
> * annotate the input to `select`: [playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1WO1AGhQ21wE84BfOAMyghDgCIABGAQwDtVOBjANYB6KOg58YAWgCOOKOWYAoRXwg94IcgBEOnOAF44AbQCMxAEzEAzMQAsxAKzEAbAF1lKtbziay8gyRYcuQAFAiKcHCyFADS6OQAXMYA5OgAbuhcya6EEVHBAGJcSSEAlAYAfHAACgwgwBgAdKKoEAA2GSGaOpyluZEYbegSJRxQUOX6VWNQjSAcYCEAyjB4XADmpYpUW57qPuR++KgBaEG4x2HReARJRnnhkZHX5HGJKemZ2f1PL0UlkyqtUYDXQzQI7U63V0HD6eQG6CGIzgIRmSQAguMOOQADxcZAgABGuAqgLgMzmC2Wq2AGy2kSouVcO0UQA)
> * use the `queryOptions` helper: [playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1WO1AGhQ21wE9iBHHKcgeTBmAgDtU4BfOAMyghBwARAAEYAQ3YSAxgGsA9FHTjpMALQ0KQgFDbpbVPBDkAIuIlwAvHADaARmIAmYgGZiAFmIBWYgDYAurp6BkbkZHRWJFi05AAUCNpwcJp0ANLo5ABctgDk6ABu6Kw5-oSJyTEAYqzZsQCUVgB8cAAK-CDAGAB0SqgQADaFscZmEnVlSRj96Kq14lBQDZbN81BdIOJgsQDKMHisAOZ12pzHwVJwxuH4HNZo0bg38Sk32TblKQxMLOzx5UmfdJZXIFIolCZJCoUaq1JbNNoCTroHoEAZDEbmcTjf5wKYzGBzBZwuCrdabHZ7YCHY5JU5lfynbRAA)

Thanks for the tips :)
I've been annotating the inputs to work around it but was shying away from `queryOptions` because these are suspense queries. I tried some workaround with `satisfies UseSuspenseQueryOptions` to narrow it down, but no luck there.

Any suggestions there? Maybe a `suspenseQueryOptions` helper?

**Comment by TkDodo** [comment]
You can pass `queryOptions` just fine to `useSuspenseQueries`

[playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgVwM4FMDKazoHYYCKy6Uw6qANCkSVAJ7UCOd9A8mDMBAXAL5wAZlAgg4AIgACMAIYFZAYwDWAeijoZCmAFoWpeuIBQhhT1TwQ9ACIzZcALxwA2gEZqAJmoBmagBZqAKzUAGwAusYmZhb0xPoONOixDAAUCIZwcHoMANLo9ABczgDk6ABu+EWhlOmZrABieIXJAJQOAHxwAAoiIMAYAHTqqBAANuXJljayzdUZGCPoWk0yUFCt9h0rUP0gMmDJmDBkeADmzYZ855HycJZJ5KjxaFg4+LSkD6lZD4VONVnsTjcAipGoZAG5ArFMoVKpg2r6BpNdYdbqiProQYUUbjSa2GQzeHzRYwZarFFwLY7PYHI7AU7nDKXaqhS6GIA)

**Comment by jtannas** [comment]
yep - was just hoping for something that'd prevent people slipping in `placeholderData`/`enabled` since that caught us pretty badly when we started transitioning to suspense

**Comment by minseong0324** [comment]
@TkDodo That's an interesting issue. Can I take a look at this?

**Comment by TkDodo** [comment]
yes please 🙏

**Comment by minseong0324** [comment]
I think it's more complicated than I expected. May I take a moment to look at it? ;)

**Comment by TmLev** [comment]
One interesting thing I discovered while combining `useSuspenseQueries` with `queryOptions` and `select` is that you can use a `satisfies` clause to make it work:

```ts
useSuspenseQueries({
  queries: [
    {
      ...someQueryOptions(),
      select: (data) => ({ // <--- `data` is `any`
        ...data,
        oldField: newValue,
      }),
    },
  ],
});

useSuspenseQueries({
  queries: [
    {
      ...someQueryOptions(),
      select: (data) => ({ // <--- `data` is inferred from `queryFn`
        ...data,
        oldField: newValue,
      }),
    } satisfies ReturnType<typeof someQueryOptions>,
  ],
});
```

Note: `select`'s return type might differ from the original `queryFn`'s return type, but you can fix that by adding an explicit return type to `select`.

**Comment by minseong0324** [comment]
I looked into this and tried adding function overloads to provide contextual typing for `select`.

It works for inline queries, but breaks explicit type parameters like `useQueries<[[string, Error, Data]]>({...})`. TypeScript can't distinguish between inferred types and explicit type parameters in overload resolution, so I couldn't find a way to support both.🥲

Let me know if you have any ideas for a different approach!

**Comment by aparajitaks** [comment]
Hi! I’d like to work on this issue.

I’ll investigate the difference in type checking for the `select` option between `useQueries` and `useQuery`, and try to align the behavior by updating the relevant TypeScript definitions. I’ll also create a minimal reproduction to verify the fix before opening a PR.

Please let me know i

**Comment by RudraKo** [comment]
Hi! I'd like to take this issue.

I'll look into the type mismatch for the `select` option on `useQueries` and work on a fix with a small reproduction and a PR.

**Comment by TkDodo** [comment]
no good solution has come up in over a year and we have a good workaround with `queryOptions`, which is the preferred abstraction anyways (we even have a lint rule for this).
